File: src/main/java/cubesrv/CubeSearch.java
package cubesrv;

import static cubesrv.CubeDefs.*;
import static cubesrv.CubesAdd.*;
import static cubesrv.CubesAddBG.*;
import cubesrv.CubeCosetsAdd;

public class CubeSearch {

    public static class CubeSearcher {
        private final int m_depthMax;
        private final CubesReprByDepthAdd m_cubesReprByDepthAdd;
        private final BGCubesReprByDepthAdd m_bgcubesReprByDepthAdd;
        private final CubeCosetsAdd m_cubeCosetsAdd;

        public CubeSearcher(int depthMax, boolean useReverse) {
            m_depthMax = depthMax;
            m_cubesReprByDepthAdd   = new CubesReprByDepthAdd(useReverse);
            m_bgcubesReprByDepthAdd = new BGCubesReprByDepthAdd(useReverse);
            m_cubeCosetsAdd         = new CubeCosetsAdd();
        }

        public void fillCubes(Responder responder) {
            m_cubesReprByDepthAdd.getReprCubes(m_depthMax, responder);
        }

        public void searchMoves(cube csearch, String mode, Responder responder) {
            String rev = m_cubesReprByDepthAdd.isUseReverse() ? " rev" : "";
            responder.message("setup: depth " + m_depthMax + rev);
            switch (mode) {
                case "q" -> SearchQuick.searchMovesQuickCatchFirst(m_cubesReprByDepthAdd,
                        m_bgcubesReprByDepthAdd, m_cubeCosetsAdd, csearch, responder);
                case "m" -> SearchQuick.searchMovesQuickMulti(m_cubesReprByDepthAdd,
                        m_bgcubesReprByDepthAdd, m_cubeCosetsAdd, csearch, responder);
                default  -> SearchOptimal.searchMovesOptimal(m_cubesReprByDepthAdd,
                        csearch, m_depthMax, responder);
            }
        }
    }
}

File: src/main/kotlin/cubesrv/CubeSearch.kt
package cubesrv

class CubeSearcher(val m_depthMax : Int, useReverse : Boolean) {
    private val m_cubesReprByDepthAdd = CubesReprByDepthAdd(useReverse)
    private val m_bgcubesReprByDepthAdd = BGCubesReprByDepthAdd(useReverse)
    private val m_cubeCosetsAdd = CubeCosetsAdd()

    fun fillCubes(responder : Responder) {
        m_cubesReprByDepthAdd.getReprCubes(m_depthMax, responder)
    }

    fun searchMoves(csearch : cube, mode : String, responder : Responder) {
        val rev = if(m_cubesReprByDepthAdd.isUseReverse()) " rev" else ""
        responder.message("setup: depth $m_depthMax$rev")
        when( mode ) {
            "q" -> searchMovesQuickCatchFirst(m_cubesReprByDepthAdd, m_bgcubesReprByDepthAdd,
                        m_cubeCosetsAdd, csearch, responder)
            "m" -> searchMovesQuickMulti(m_cubesReprByDepthAdd, m_bgcubesReprByDepthAdd,
                        m_cubeCosetsAdd, csearch, responder)
            else -> searchMovesOptimal(m_cubesReprByDepthAdd, csearch, m_depthMax, responder)
        }
    }
}

